gnocchi/storage/__init__.py

```diff
@@ -365,35 +378,37 @@ def _update_metric_splits(self, metrics_keys_aggregations_splits):
                                                   oldest_mutable_timestamp)
         """
         metrics_splits_to_store = {}
+        keys_to_get = collections.defaultdict(
+            lambda: collections.defaultdict(list))
+        splits_to_rewrite = collections.defaultdict(
+            lambda: collections.defaultdict(list))
 
         for metric, (keys_and_aggregations_and_splits,
                      oldest_mutable_timestamp) in six.iteritems(
                          metrics_keys_aggregations_splits):
-            keys_to_rewrite = []
-            splits_to_rewrite = []
             for (key, aggregation), split in six.iteritems(
                     keys_and_aggregations_and_splits):
                 # NOTE(jd) We write the full split only if the driver works
                 # that way (self.WRITE_FULL) or if the oldest_mutable_timestamp
                 # is out of range.
-                write_full = (
-                    self.WRITE_FULL or next(key) <= oldest_mutable_timestamp
-                )
-                if write_full:
-                    keys_to_rewrite.append(key)
-                    splits_to_rewrite.append(split)
-
-            # Update the splits that were passed as argument with the data
-            # already stored in the case that we need to rewrite them fully.
-            # First, fetch all those existing splits.
-            existing_data = self._get_splits_and_unserialize(
-                metric, [(key, split.aggregation)
-                         for key, split
-                         in six.moves.zip(keys_to_rewrite, splits_to_rewrite)])
-
-            for key, split, existing in six.moves.zip(
-                    keys_to_rewrite, splits_to_rewrite, existing_data):
-                if existing:
+                if self.WRITE_FULL or next(key) <= oldest_mutable_timestamp:
+                    # Update the splits that were passed as argument with the
+                    # data already stored in the case that we need to rewrite
+                    # them fully. First, fetch all those existing splits.
+                    keys_to_get[metric][aggregation].append(key)
+                    splits_to_rewrite[metric][aggregation].append(split)
+
+        existing_data = self._get_splits_and_unserialize(keys_to_get)
+
+        for metric, (keys_and_aggregations_and_splits,
+                     oldest_mutable_timestamp) in six.iteritems(
+                         metrics_keys_aggregations_splits):
+            for aggregation, existing_list in six.iteritems(
```

Comment:
:/ we're doing a lot of extra looping to build format. hope it doesn't affect the unbatched routes.